src/compiler/checker.ts

```diff
@@ -36505,8 +36507,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         else if (isAssertionExpression(expr) && !isConstTypeReference(expr.type)) {
             return getTypeFromTypeNode((expr as TypeAssertion).type);
         }
-        else if (node.kind === SyntaxKind.NumericLiteral || node.kind === SyntaxKind.StringLiteral ||
-            node.kind === SyntaxKind.TrueKeyword || node.kind === SyntaxKind.FalseKeyword) {
+        else if (isLiteralExpression(node) || isBooleanLiteral(node)) {
             return checkExpression(node);

```

Comment:
Note that the test suite passes even if we **remove** this `else if` block entirely - leading me to believe that this is merely an optimization for those types here (as it's not an unused code ⚠️ ).

The main improvement here is to just cover every literal type - including bigints, regexps and "no substitution template literals".

Feel free to close the PR if you don't like this change.